lupa/_lupa.pyx

```diff
@@ -1631,8 +1712,10 @@ cdef run_lua(LuaRuntime runtime, bytes lua_code, tuple args):
     try:
         check_lua_stack(L, 1)
         if lua.luaL_loadbuffer(L, lua_code, len(lua_code), '<python>'):
-            raise LuaSyntaxError(build_lua_error_message(
-                runtime, L, err_message=u"error loading code: %s"))
+            error = build_lua_error_message(runtime, L)
+            if error == "not enough memory":
+                raise LuaMemoryError(error)
```

Comment:
It seems that "not enough memory" isn't always the (complete) error message here. We need either a safer or more relaxed way to detect memory errors.
```
ERROR: test_not_enough_memory (lupa.tests.TestMaxMemory_lua53)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/lupa/lupa/lupa/tests/test.py", line 3019, in test_not_enough_memory
    self.assertRaises(lupa.LuaMemoryError, self.lua.eval, "('a'):rep(50000)")
  File "/opt/hostedtoolcache/Python/2.7.18/x64/lib/python2.7/unittest/case.py", line 473, in assertRaises
    callableObj(*args, **kwargs)
  File "lupa/lua53.pyx", line 395, in lupa.lua53.LuaRuntime.eval
    return run_lua(self, b'return ' + lua_code, args)
  File "lupa/lua53.pyx", line 1721, in lupa.lua53.run_lua
    return call_lua(runtime, L, args)
  File "lupa/lua53.pyx", line 1735, in lupa.lua53.call_lua
    return execute_lua_call(runtime, L, len(args))
  File "lupa/lua53.pyx", line 1798, in lupa.lua53.execute_lua_call
    raise_lua_error(runtime, L, result_status)
  File "lupa/lua53.pyx", line 1650, in lupa.lua53.raise_lua_error
    raise LuaError(build_lua_error_message(runtime, L))
LuaError: [string "<python>"]:1: not enough memory for buffer allocation
stack traceback:
	[string "<python>"]:1: in main chunk
	[C]: in function 'string.rep'
```

Reply:
```suggestion
            if error.startswith("not enough memory"):
                raise LuaMemoryError(error)
```

Reply:
I think the error message above starts with this: "[string "<python>"]:1: not enough memory for buffer allocation"

Reply:
(You may have to merge the master branch into yours, as the CI runner does.)

Reply:
Isn't there an error code somewhere in Lua that we could look at instead? And that would allow us to safely detect a specific type of error?

Reply:
Actually, we already do look at the error codes, in `raise_lua_error()`. When we see a `LUA_ERRMEM`, we raise a `MemoryError`.
https://github.com/scoder/lupa/blob/c92dc8208320ca0f0757e2feac5347b26cf92916/lupa/_lupa.pyx#L1557-L1563

Reply:
And I just noticed that we previously raised a standard Python `MemoryError` for memory allocation failures. Switching to a `LuaMemoryError` that does not inherit from `MemoryError` is an incompatible change. I don't see a reason to do that. We can just continue to raise `MemoryError()` (without message), I think.

Reply:
Python's MemoryError does not inherit from Lupa's LuaError, so try/catchs with only LuaError may not catch it.
I'd prefer changing LuaMemoryError to inherit from MemoryError.

(also thanks for github for not loading comments, so I almost missed all your comments when I resolved this)